cmd/launcher/doctor.go

```diff
@@ -0,0 +1,531 @@
+		RootPEM:                            *flRootPEM,
+		Transport:                          *flTransport,
+		UpdateChannel:                      updateChannel,
+		UpdateDirectory:                    *flUpdateDirectory,
+	}
+
+	return opts, nil
+}
+
+func (c *checkup) run() {
+	if c.check == nil {
+	}
+
+	cRunning.Printf("\nRunning checkup: ")
+	cHeader.Printf("%s\n", c.name)
+
+	result, err := c.check()
+	if err != nil {
+		info(result)
+		bad(err)
+		fail("𐄂 Checkup failed!")
+	} else {
+		good(result)
+		pass("✔ Checkup passed!")
+	}
+}
+
+func checkupPlatform(os string) (string, error) {
+	if slices.Contains([]string{"windows", "darwin", "linux"}, os) {
+		return fmt.Sprintf("Platform: %s", os), nil
+	}
+	return "", fmt.Errorf("Unsupported platform: %s", os)
+}
+
+type launcherFile struct {
+	name  string
+	found bool
+}
+
+func checkupRootDir(filepaths []string) (string, error) {
+	importantFiles := []*launcherFile{
+		{
+			name: "debug.json",
+		},
+		{
+			name: "launcher.db",
+		},
+		{
+			name: "osquery.db",
+		},
+	}
+
+	if filepaths != nil && len(filepaths) > 0 {
+		for _, fp := range filepaths {
+			for _, f := range importantFiles {
+				if filepath.Base(fp) == f.name {
+					f.found = true
+				}
+			}
+		}
+	}
+
+	var failures int
+	for _, f := range importantFiles {
+		if f.found {
+			good(f.name)
+		} else {
+			bad(f.name)
+			failures = failures + 1
+		}
+	}
+
+	if failures == 0 {
+		return "Root directory files found", nil
+	}
+
+	return "", fmt.Errorf("%d root directory files not found", failures)
+}
+
+func checkupConnectivity(logger log.Logger, k types.Knapsack) (string, error) {
+	var failures int
+	checkpointer := checkpoint.New(logger, k)
+	connections := checkpointer.Connections()
+	for k, v := range connections {
+		if v == "successful tcp connection" {
+			good(fmt.Sprintf("%s - %s", k, v))
+		} else {
+			bad(fmt.Sprintf("%s - %s", k, v))
+			failures = failures + 1
+		}
+	}
+
+	ipLookups := checkpointer.IpLookups()
+	for k, v := range ipLookups {
+		valStrSlice, ok := v.([]string)
+		if ok && len(valStrSlice) > 0 {
+			good(fmt.Sprintf("%s - %s", k, valStrSlice))
+		} else {
+			bad(fmt.Sprintf("%s - %s", k, valStrSlice))
+			failures = failures + 1
+		}
+	}
+
+	notaryVersions, err := checkpointer.NotaryVersions()
+	if err != nil {
+		bad(fmt.Errorf("could not fetch notary versions: %w", err))
+		failures = failures + 1
+	}
+
+	for k, v := range notaryVersions {
+		if _, err := strconv.ParseInt(v, 10, 32); err == nil {
+			good(fmt.Sprintf("%s - %s", k, v))
+		} else {
+			bad(fmt.Sprintf("%s - %s", k, v))
+			failures = failures + 1
+		}
+	}
+
+	if failures == 0 {
+		return "Successfully communicated with Kolide", nil
+	}
+
+	return "", fmt.Errorf("%d failures encountered while attempting communication with Kolide", failures)
+}
+
+func checkupVersion(v version.Info) (string, error) {
+	info(fmt.Sprintf("Current version: %s", v.Version))
+	// TODO: Query TUF for latest available version for this launcher instance
```

Comment:
Not yet sure if this is easy to do? Have to factor in current version, update channel, etc.

Reply:
I think you could do this fairly simply once https://github.com/kolide/launcher/pull/1185 is merged in:

1. Export this function so you can init a metadata client for yourself here: https://github.com/kolide/launcher/blob/main/pkg/autoupdate/tuf/autoupdate.go#L117
1. Export `findRelease`: https://github.com/kolide/launcher/blob/0ae45a14810c2d3be77d786840def2bfe997cad0/pkg/autoupdate/tuf/autoupdate.go#L328
1. Export `versionFromTarget` or implement it yourself: https://github.com/kolide/launcher/blob/main/pkg/autoupdate/tuf/library_manager.go

Then something like:

```
// Make a temp TUF directory so that we don't change anything about our currently in-use local TUF repo
tempTufRepoDir, _ := agent.MkdirTemp("temp-tuf")

// Get a metadata client -- this will be initialized with the root metadata we distribute, so we can trust the downloads it performs
tufClient, _ := tuf.InitMetadataClient(tempTufRepoDir, k.TufServerURL(), http.DefaultClient)

// Talks to TUF server to fetch all targets
tufClient.Update()

// Finds the release target (release.json), finds the target it points to, gets the version from that target
targets, _ := tufClient.Targets()
releaseVersion, _, _ := tuf.FindRelease("launcher", targets, k.UpdateChannel())
fmt.Println(tuf.VersionFromTarget("launcher", releaseVersion))
```

Reply:
I'm also going to have to do something very similar for the packaging/installer part of https://github.com/kolide/launcher/issues/1149 (make sure we download the correct version for the given channel, once we can't just download `launcher-stable.tar.gz` any more). So maybe I could just add an exported function to the `tuf` package that does this.

Reply:
Added 🙂 https://github.com/kolide/launcher/pull/1185/commits/b612e8158037119191261db6e769200219b05bdf